### PR TITLE
box: introduce interface for tuple constraints

### PR DESCRIPTION
**Pull request by ImeevMA**
This patch introduces drop_constraint(), drop_foreign_key(), create_constraint() and create_foreign_key() methods for space in Lua.

Needed for #6986

### CODE REVIEW

File: src/box/lua/schema.lua
Comment (Gerold103) [inline src/box/lua/schema.lua:2731]: Don't need `;`.

File: src/box/lua/schema.lua
Comment (Gerold103) [inline src/box/lua/schema.lua:2732]: Don't need `()`.

File: src/box/lua/schema.lua
Comment (Gerold103) [inline src/box/lua/schema.lua:2737]: I see other usages of this function employ `''` (single quotes). I suggest to do the same here and in `normalize_foreign_key()`.

File: src/box/schema.cc
Comment (Gerold103) [inline src/box/schema.cc:497]: Can be `static`. Doesn't seem to be used outside of this file.

File: src/box/schema.cc
Comment (Gerold103) [inline src/box/schema.cc:512]: Why is it called `str` if it is not a string?

File: src/box/schema.cc
Comment (Gerold103) [inline src/box/schema.cc:567]: Instead of all those tricky decodings could try to re-use `space_def_new_from_tuple()`. Although you would need to implement `space_def_to_tuple()` or `space_def_to_msgpack()`. Then you would decode a `space_def`, do the needed changes without the msgpack tricks, encode it back, and use `box_replace()` like now. Up to you if you want to try.

File: src/box/schema.cc
Comment (Gerold103) [inline src/box/schema.cc:498]: Having this new API in schema.h looks odd. Mostly because this file had nothing to do with the storage layer before (and this is why it didn't need box.h). All was in box.h/.cc like `box_insert()`, `box_sequence_next()`, `boxk()` and all the similar functions. I would consider moving it all there. Up to you.

File: src/box/schema.cc
Comment (Gerold103) [inline src/box/schema.cc:594]: You might think about `box_update()` with `#` operation. It supports JSON paths, so you can build msgpack operations like `{{'#', 'path.to.constraint.def'}}`. In case it won't fail on a not existing field (I don't remember if it does), then on success you can compare old and new tuples using `memcmp`. If they are equal, it means the update didn't do anything and the constraint wasn't found. `box_update()` could simplify this code a lot.

Even to build the operations msgpack you could use `mp_vformat()` so as not to encode anything manually.

Same for constraints creation.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch!

File: extra/exports
Comment (ImeevMA) [reply extra/exports]: Thank you for the review! I have dropped these functions from `exports` as you suggested below.

File: src/box/lua/schema.lua
Comment (ImeevMA) [reply src/box/lua/schema.lua]: Thanks, fixed!

File: src/box/lua/schema.lua
Comment (ImeevMA) [reply src/box/lua/schema.lua]: Fixed.

File: src/box/lua/schema.lua
Comment (ImeevMA) [reply src/box/lua/schema.lua]: Fixed.

File: src/box/lua/schema.lua
Comment (ImeevMA) [reply src/box/lua/schema.lua]: Fixed.

File: src/box/lua/schema.lua
Comment (ImeevMA) [reply src/box/lua/schema.lua]: Fixed.

File: src/box/schema.cc
Comment (ImeevMA) [reply src/box/schema.cc]: Fixed.

File: src/box/schema.cc
Comment (ImeevMA) [reply src/box/schema.cc]: I used `str` as a generic name here, but now I think `buf` or `pos` would be better. In any case, I dropped this part.

File: src/box/schema.cc
Comment (ImeevMA) [reply src/box/schema.cc]: I used your suggestion about `box_update()`.

File: src/box/schema.cc
Comment (ImeevMA) [reply src/box/schema.cc]: Thanks, moved.

File: src/box/schema.cc
Comment (ImeevMA) [reply src/box/schema.cc]: Thank you for the suggestion! The use of JSON paths helped a lot here - now the code is much simpler and a little shorter.

File: src/box/box.cc
Comment (Gerold103) [inline src/box/box.cc:4346]: Just in case you missed my first comment, you can also use `mp_vformat()`, it might be significantly simpler than manual encoding and it is not like we are running for perf here. Up to you.

**Review by Gerold103 — COMMENTED** [review]
Hi! Thanks for the fixes! Now there is a new problem though, about `box_update()`, which I didn't think of before:
```Lua
box.cfg{}
body = "function(x) return true end"
box.schema.func.create('ck1', {is_deterministic = true, body = body})
func_id = box.func.ck1.id
fk0 = {one = {field = {a = 'a'}}, two = {field = {b = 'b'}}}
ck0 = {three = 'ck1', four = 'ck1'}
fk1 = {five = {field = 'a'}, six = {field = 'b'}}
ck1 = {seven = 'ck1', eight = 'ck1'}
fmt = {{'a', 'integer'}, {'b', 'integer'}}
fmt[1].constraint = ck1
fmt[2].foreign_key = fk1
def = {format = fmt, foreign_key = fk0, constraint = ck0}
s = box.schema.space.create('a', def)
t = box.space._space:get{s.id}
t = t:totable()
t[6].foreign_key = setmetatable({}, {__serialize = 'map'})
t[6].constraint = t[6].foreign_key
box.space._space:replace(t)
box.space.a:create_constraint('four', 'ck1')
```
In your code when you create a constraint, you assume that the corresponding map does not exist in case there were no constraints before. But it can exist as an empty map. In that case `!` operation would fail:
```
---
- error: 'Field ''flags.constraint'' UPDATE error: the key exists already'
...
```
Not sure what would be the best way to fix this. We certainly shouldn't make a `box_update()` like now and redo it if it fails as a duplicate. You could try to get the tuple first, then check if it has the needed map using `tuple_field_raw_by_full_path()` (so as not to bother with decoding things manually, or you can decode them manually too if want). If it does, then you insert into it, or otherwise insert the map itself too.

After fixing it I suggest to add tests on this. Deletion of constraints looks fine right now, the problem is just with their creation AFAICS.

File: src/box/box.cc
Comment (ImeevMA) [reply src/box/box.cc]: Thank you for the review! I fixed the above bug. Also I refactored the code using `mp_vformat()`. I really missed that part of the comment.

File: src/box/box.cc
Comment (Gerold103) [inline src/box/box.cc:4306]: The same is done in `boxk()`. I suggest to use it there too. Should be easy given that it is in the same file (box.cc).

File: src/box/box.cc
Comment (Gerold103) [inline src/box/box.cc:4367]: `%d` -> `%u`. Same in `constraint_create()`.

File: src/box/box.cc
Comment (Gerold103) [inline src/box/box.cc:4488]: I suggest to assert here that the type is either FUNC or FKEY.

File: src/box/box.cc
Comment (Gerold103) [inline src/box/box.cc:4505]: Need to truncate the region here. Below there were done some `mp_format_on_region()` calls.

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the fixes! Some last comments.

File: src/box/box.cc
Comment (ImeevMA) [reply src/box/box.cc]: Thanks for the review! I tried to use `mp_format_on_region()` but it's quite problematic as we can't call one variadic function from another. We can use `ap_list` instead of `...` for `mp_format_on_region()`, but we have to create a new function that creates `ap_list` for `constraint_create()` and `constraint_drop()`. Not sure if this is needed.

I added size as a new output argument to mp_format_on_region().

File: src/box/box.cc
Comment (ImeevMA) [reply src/box/box.cc]: Fixed.

File: src/box/box.cc
Comment (ImeevMA) [reply src/box/box.cc:4493]: Thanks, added.

File: src/box/box.cc
Comment (ImeevMA) [reply src/box/box.cc:4513]: Thanks, fixed.

File: src/box/box.cc
Comment (Gerold103) [reply src/box/box.cc]: Try this:
<details><summary>Diff</summary>

```Diff
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -2312,6 +2312,41 @@ box_set_txn_isolation(void)
 
 /* }}} configuration bindings */
 
+/**
+ * Encode a sequence of values according to format string into a buffer
+ * allocated on region.
+ */
+static const char *
+mp_vformat_on_region(struct region *region, size_t *size, const char *format,
+		     va_list src)
+{
+	va_list ap;
+	va_copy(ap, src);
+	size_t buf_size = mp_vformat(NULL, 0, format, ap);
+	char *buf = (char *)region_alloc(region, buf_size);
+	va_end(ap);
+	if (buf == NULL) {
+		diag_set(OutOfMemory, buf_size, "region_alloc", "buf");
+		return buf;
+	}
+	va_copy(ap, src);
+	mp_vformat(buf, buf_size, format, ap);
+	va_end(ap);
+	*size = buf_size;
+	return buf;
+}
+
+static const char *
+mp_format_on_region(struct region *region, size_t *size, const char *format,
+		    ...)
+{
+	va_list ap;
+	va_start(ap, format);
+	const char *res = mp_vformat_on_region(region, size, format, ap);
+	va_end(ap);
+	return res;
+}
+
 /**
  * Execute a request against a given space id with
  * a variable-argument tuple described in format.
@@ -2335,17 +2370,13 @@ boxk(int type, uint32_t space_id, const char *format, ...)
 	request.type = type;
 	request.space_id = space_id;
 	va_start(ap, format);
-	size_t buf_size = mp_vformat(NULL, 0, format, ap);
-	char *buf = (char *)region_alloc(&fiber()->gc, buf_size);
+	struct region *region = &fiber()->gc;
+	size_t size = 0;
+	const char *data = mp_vformat_on_region(region, &size, format, ap);
 	va_end(ap);
-	if (buf == NULL)
+	if (data == NULL)
 		return -1;
-	va_start(ap, format);
-	if (mp_vformat(buf, buf_size, format, ap) != buf_size)
-		assert(0);
-	va_end(ap);
-	const char *data = buf;
-	const char *data_end = buf + buf_size;
+	const char *data_end = data + size;
 	switch (type) {
 	case IPROTO_INSERT:
 	case IPROTO_REPLACE:
@@ -4298,30 +4329,6 @@ box_read_ffi_enable(void)
 		box_read_ffi_is_disabled = false;
 }
 
-/**
- * Encode a sequence of values according to format string into a buffer
- * allocated on region.
- */
-static const char *
-mp_format_on_region(struct region *region, uint32_t *size, const char *format,
-		    ...)
-{
-	va_list ap;
-	va_start(ap, format);
-	size_t buf_size = mp_vformat(NULL, 0, format, ap);
-	char *buf = (char *)region_alloc(region, buf_size);
-	va_end(ap);
-	if (buf == NULL) {
-		diag_set(OutOfMemory, buf_size, "region_alloc", "buf");
-		return buf;
-	}
-	va_start(ap, format);
-	mp_vformat(buf, buf_size, format, ap);
-	va_end(ap);
-	*size = buf_size;
-	return buf;
-}
-
 /**
  * Drop the constraint or foreign key. If fieldno is equal to UINT32_MAX then
  * the tuple constraint or tuple foreign key is dropped, otherwise the field
@@ -4378,7 +4385,7 @@ constraint_drop(uint32_t space_id, const char *name, uint32_t name_len,
 		snprintf(buf, size, "%s.%*s", path, name_len, name);
 		path = buf;
 	}
-	uint32_t ops_size;
+	size_t ops_size;
 	const char *ops = mp_format_on_region(region, &ops_size, "[[%s%s%u]]",
 					      "#", path, 1);
 	const char *end = ops + ops_size;
@@ -4485,7 +4492,7 @@ constraint_create(uint32_t space_id, const char *name, uint32_t name_len,
 	struct region *region = &fiber()->gc;
 	uint32_t used = region_used(region);
 	const char *val;
-	uint32_t val_size;
+	size_t val_size;
 	assert(type == CONSTR_FUNC || type == CONSTR_FKEY);
 	if (type == CONSTR_FUNC) {
 		val = mp_format_on_region(region, &val_size, "%u",
@@ -4502,7 +4509,7 @@ constraint_create(uint32_t space_id, const char *name, uint32_t name_len,
 	if (val == NULL)
 		return -1;
 	const char *ops;
-	uint32_t ops_size;
+	size_t ops_size;
 	if (is_empty) {
```
</details>

**Review by Gerold103 — APPROVED** [review]
Thanks for the fixes! Technically looks good. See one remaining comment among the old ones where I've shown how to reuse 'mpformat on region'.

File: src/box/box.cc
Comment (ImeevMA) [reply src/box/box.cc]: Thank you for the diff! Applied.

File: changelogs/unreleased/gh-6986-introduce-constraint-interface.md
Comment (locker) [inline changelogs/unreleased/gh-6986-introduce-constraint-interface.md:4]: We shouldn't modify or extend the existing API without an approval from the PM team. Anyway, I don't understand *why* you need these new Lua methods. The idea was to introduce the C API that could be used by SQL. Ideally, we should rewrite the *existing* Lua API using it if possible, not introduce *new* API.

**Review by locker — COMMENTED** [review]
1. Please refrain from setting the `full-ci` label - the label is supposed to be set before merge, when all necessary approvals are obtained. If you need to test your PR on a certain platform (e.g. it modifies the platform-specific code)
2. Why do you need two PRs - this is confusing. Should you close this one and leave only #7651?

File: changelogs/unreleased/gh-6986-introduce-constraint-interface.md
Comment (ImeevMA) [reply changelogs/unreleased/gh-6986-introduce-constraint-interface.md:4]: I convert this PR to draft for now.

**Comment by ImeevMA** [comment]
It doesn't look like this PR will be continued anytime soon. Closing.

**Pull request closed without merging**
